src/ufuncs.c

```diff
@@ -382,6 +382,163 @@ static void YY_b_p_func(char** args, npy_intp* dimensions, npy_intp* steps, void
 }
 static PyUFuncGenericFunction YY_b_p_funcs[1] = {&YY_b_p_func};
 
+/* Define the geom, X, Y -> bool functions (Ydd_b) prepared */
+static char GEOSContainsXY(void* context, void* g1, void* pg1, double x, double y) {
+  GEOSGeometry *geom = NULL;
+  const GEOSPreparedGeometry* prepared_geom_tmp = NULL;
+  char ret;
+
+#if !GEOS_SINCE_3_12_0
+  geom = create_point(context, x, y);
+  if (geom == NULL) {
+    return 2;
+  }
+#endif
+
+  char destroy_prepared = 0;
+  if (pg1 == NULL) {
+    prepared_geom_tmp = GEOSPrepare_r(context, g1);
+    if (prepared_geom_tmp == NULL) {
+      return 2;
+    }
+    destroy_prepared = 1;
+  } else {
+    prepared_geom_tmp = pg1;
+  }
+
+#if GEOS_SINCE_3_12_0
+  ret = GEOSPreparedContainsXY_r(context, prepared_geom_tmp, x, y);
+#else
+  ret = GEOSPreparedContains_r(context, prepared_geom_tmp, geom);
+#endif
+  if (destroy_prepared) {
+    GEOSPreparedGeom_destroy_r(context, prepared_geom_tmp);
+  }
+#if !GEOS_SINCE_3_12_0
+  GEOSGeom_destroy_r(context, geom);
+#endif
+  return ret;
+}
+static void* contains_xy_data[1] = {GEOSContainsXY};
+
+static char GEOSContainsProperlyXY(void* context, void* g1, void* pg1, double x, double y) {
```

Comment:
Those functions (for contains, contains_properly, and intersects) are a bit repetitive. Another option could also be to move this into the actual `Ydd_b_p_func`, and inject `GEOSPreparedContainsXY_r` and `GEOSPreparedContains_r` as `data` (similarly how `YY_b_p_func` gets the non-prepared and prepared version of the predicate).

(however, in case we would drop the on the fly preparation, and assume this is already done in advance, that would make those functions a lot shorter)